onnx/shape_inference/implementation.cc

```diff
@@ -161,5 +179,98 @@ void InferShapes(
   }
 }
 
+void InferShapeForFunctionNode(
+    const FunctionProto& func,
+    const ISchemaRegistry* schema_registry,
+    InferenceContext& ctx) {
+  int domain_version = (int)func.since_version();
+  GraphProto g;
+  // Get a temproary tensor-shape map
+  std::unordered_map<std::string, TypeProto*> temp_valueTypesByName;
+  std::vector<TypeProto> temp_types_cache;
+  for (int i = 0; i < func.input_size(); ++i) {
+    temp_types_cache.push_back(*ctx.getInputType(i));
+    temp_valueTypesByName[func.input().Get(i)] = &temp_types_cache.back();
+  }
+  // Get a temproary initial value map
+  std::unordered_map<std::string, const TensorProto*>
+      temp_initializersByName;
+  for (int i = 0; i < (const int)(ctx.getNumInputs()); ++i) {
+    if (ctx.getInputData(i) != nullptr && i < func.input_size()) {
+      temp_initializersByName[func.input().Get(i)] = ctx.getInputData(i);
+    }
+  }
+  std::unordered_map<std::string, const AttributeProto*> attr_map;
+  for (auto& attr : func.attribute()) {
+    if (ctx.getAttribute(attr) != nullptr) {
+      attr_map[attr] = ctx.getAttribute(attr);
+    }
+  }
+
+  for (auto& n : func.node()) {
+    const auto schema =
+        schema_registry->GetSchema(n.op_type(), domain_version, n.domain());
+    if (!schema) {
+      return;
+    }
+    NodeProto copy_n(n);
+    // Add attribute information into the temproary node
+    copy_n.clear_attribute();
+    for (auto attr : n.attribute()) {
+      if (attr.has_ref_attr_name()) {
+        if (attr_map.count(attr.ref_attr_name())) {
+          copy_n.add_attribute()->CopyFrom(*attr_map[attr.ref_attr_name()]);
+        }
+      } else {
+        copy_n.add_attribute()->CopyFrom(attr);
+      }
+    }
+
+    InferenceContextImpl temp_ctx(
```

Comment:
We actually need to have a new context for every node. It will pick out the existing shape/type info from the temp_valuetypesbyname in its constructor.